### ISSUE TITLE
proposal: metadata service -> identity service

### ISSUE THREAD

**Issue by jonboulle** [appc/spec#270]
The [metadata service](https://github.com/appc/spec/blob/2740753821dff142c56f2d241b0e7386a4bdaee5/SPEC.md#app-container-metadata-service) is a part of the spec that often causes much confusion about its purpose and utility. In discussions yesterday about how to make things clearer, one option that came up is to rename it to _Identity Service_ to emphasise the idea of providing a unique identity to each container and allowing containers to verify that identity across a federated instance of the service. /cc @philips @eyakubovich

**Comment by eyakubovich** [comment]
Having not been in discussions where the purpose of metadata service has been a source of confusion, do the users perceive the value of metadata (uuid, manifests, annotations) as marginal or do they not understand how to retrieve it? There has to be an accepted way to feed arbitrary key=value to the apps. If this is through annotations, I would like not to de-emphasize this purpose through a rename. If it's through env vars, then I'm +1 on the rename.

**Comment by jonboulle** [comment]
> do users perceive the value of metadata (uuid, manifests, annotations) as marginal

That is my experience, yes. For argument's sake, let's consider 12 factor apps, which are idiomatically configured through [environment variables](http://12factor.net/config) - since they should be [stateless](http://12factor.net/processes) and [disposable](http://12factor.net/disposability) there is arguably no need for them to have a requirement on this metadata.

**Comment by jonboulle** [comment]
(Playing devil's advocate somewhat - I _personally_ see value in the service, but am not finding it a particularly common view.)

**Comment by eyakubovich** [comment]
Let's also explore the flip side. Do the users see the value in the identity part of the service? Do we know of anyone doing anything interesting with it yet?

**Comment by jonboulle** [comment]
Weird, thought I had written some notes on that already..

@eyakubovich obviously there is a bit of a chicken and egg issue here when talking about "users". My take from talking about this to a number of interested parties is that there is an unfortunate catch-22 with the identity service, in that its value tends to be most readily recognised/demanded in large and complex environments, but such environments typically already have their own bespoke implementation of a service providing the same features/value, and have so much existing investment in those systems that it would be extremely unlikely they would migrate to the appc metadata service. (It would be really great if some of said environments published whitepapers about this stuff, then people would be clamouring for it ;-)

**Comment by eyakubovich** [comment]
@jonboulle you have mentioned this before and I understand the challenge. But it also means that we don't have validation yet that this will be embraced. I just want to few folks to comment here (or somewhere) on how they think they'll utilize the crypto-identity in their apps.

**Comment by jonboulle** [comment]
I agree. More voices would be useful, both from other spec implementers and
general users. /cc @cdaylward @mpasternacki @krobertson

**Comment by tnachen** [comment]
I recommend the name change if it's merely about identity and authenticating. I assumed it was also providing metadata of images or pods when I first heard about the service from the spec, then realized it wasn't.

**Comment by jonboulle** [comment]
@tnachen it _does_ provide the metadata of images or pods, is that not clear?
https://github.com/appc/spec/blob/master/SPEC.md#pod-metadata
https://github.com/appc/spec/blob/master/SPEC.md#app-metadata

**Comment by tnachen** [comment]
I see, I saw the first line where it was mentioned in the beginning:

"The Metadata Service defines how apps within pods can introspect and get a cryptographically verifiable identity from the execution environment."

And also I asked during our meeting about it so I assumed it didn't provide the metadata. Seems like in this case calling it metadata service probably is more generic.

**Comment by jonboulle** [comment]
You are right, the wording is slightly ambivalent throughout (e.g. [here](https://github.com/appc/spec/blob/master/SPEC.md#app-container-metadata-service) we talk about metadata only instead of mentioning the identity). Let's clear that up as part of this issue.

**Comment by krobertson** [comment]
I agree that metadata can be a conflated term... identity is likely better, but may still mean different things to different people. The best name would have to be about its intended purpose.

We've seen a couple different use cases for a [insert-name] service able to return app information:

1. Identity, essentially ephemeral key type info so it can talk to others or validate others who are talking to it.
2. Basic image specification data
3. Runtime environment data

Being able to provide some cryptographic identiy to a process would be very valuable. Being able to have multiple apps within a cluster cross communicating and validating identity through the platform has a lot of power. It takes authentication out of the apps and leverages the trust in the platform. The platform that is executing you is the best authority for whether your neighbor really is your neighbor.

Some identity information can be useful to pass in through environment, like the pod's UUID, its name, and some non-sensitive information for a quick reference. But an app may want to know additional information about itself from its image manifest, such as how much memory was it allocated. A java app can use this to set its heap, for instance. But putting that information in env vars gets bloated, so something like an API to query is very useful. Today, we are doing that through runtime templates that get written to, but it isn't ideal since it is only accessible on startup and not after the fact.

And finally environmental information, beyond just the specification but information about where it is running. So being able to say this process is in us-west-1a, or on-prem in rack C1R2U40. It could also act as a broker to a configuration service, like etcd/zookeeper, where rather than needing the app to query etcd, the platform can and provide it the same metadata. "Your DB is over there."

For different types of metadata, a metadata service may be fine, since features provided beyond the baseline spec may be a place where appc implementations differentiate.

Personally think either is suitable and can see our use cases still fitting in with identity.

**Comment by jonboulle** [comment]
@krobertson that is really valuable feedback, thanks!
